src/init.cpp

```diff
@@ -1807,6 +1814,21 @@ bool AppInitMain(NodeContext& node)
         return false;
     }
 
+    // set asmap if supplied
+    if (gArgs.IsArgSet("-asmap")) {
+        const fs::path asmap_path = GetDataDir() / gArgs.GetArg("-asmap", DEFAULT_ASMAP_FILENAME);
```

Comment:
This appears to be a bug. Suggested fix:
```diff
-        const fs::path asmap_path = GetDataDir() / gArgs.GetArg("-asmap", DEFAULT_ASMAP_FILENAME);
+        const fs::path asmap_path = GetDataDir() / DEFAULT_ASMAP_FILENAME;
```

Before this fix, passing `-asmap` with no file specified returns:
`Error: Could not parse specified asmap: '"/home/jon/.bitcoin"'`

After, it correctly errors by looking for the default file defined in [line 104](https://github.com/bitcoin/bitcoin/pull/16702/files#diff-c865a8939105e6350a50af02766291b7R104) above:
`Error: Could not parse specified asmap: '"/home/jon/.bitcoin/ip_asn.map"'`

Reply:
I think what you want here is:
```diff
-        const fs::path asmap_path = GetDataDir() / gArgs.GetArg("-asmap", DEFAULT_ASMAP_FILENAME);
+        std::string asmap_file = gArgs.GetArg("-asmap", "");
+        if (asmap_file.empty()) {
+            asmap_file = DEFAULT_ASMAP_FILENAME;
+        }
+        const fs::path asmap_path = GetDataDir() / asmap_file;
```
This provides the behavior of defaulting to `DEFAULT_ASMAP_FILENAME` if the user passes `-asmap` without a filename.

Reply:
Weird. You're handling a special case when somebody passes -asmap="".
What if somebody passes -asmap=" " or asmap="/"? These are conditions, which are equally possible in users stupidity. If a user don't want asmap, they ignore the flag.
I don't think protecting against a specific case of empty string is reasonable. Other opinions?

Reply:
I could be misunderstanding, but in what case is `DEFAULT_ASMAP_FILENAME` otherwise?

Reply:
This handles the case for passing `-asmap`. FWIW, I have been testing this PR rebased to master.

Reply:
If I understand the intended interface in this PR:

1) `bitcoind`: use /16 bucketing
2) `bitcoind -asmap`: use default asmap file in the datadir
3) `bitcoind -asmap=<path-to-file>`: use specifed asmap

My suggestion handles case 2.

I could be confused, but I don't see in what case `DEFAULT_ASMAP_FILENAME` would be used otherwise. AFAICT, behind the `if (gArgs.IsArgSet("-asmap"))` conditional, `DEFAULT_ASMAP_FILENAME` in `gArgs.GetArg("-asmap", DEFAULT_ASMAP_FILENAME)` will never be called.

Reply:
I think the line was just supposed to be
```suggestion
        const fs::path asmap_path = GetDataDir() / gArgs.GetArg("-asmap", DEFAULT_ASMAP_FILENAME);
```
const fs::path asmap_path = GetDataDir() / DEFAULT_ASMAP_FILENAME;
```

This fixed the behavior for me:

With current code starting `bitcoind -asmap` (with no default file present):
```
2019-12-15T21:40:36Z Opened asmap file (864 bytes) from disk.
2019-12-15T21:40:36Z

************************
EXCEPTION: NSt3__18ios_base7failureE
CAutoFile::read: fread failed: unspecified iostream_category error
bitcoin in AppInit()



************************
EXCEPTION: NSt3__18ios_base7failureE
CAutoFile::read: fread failed: unspecified iostream_category error
bitcoin in AppInit()
```

With changed line:
```
2019-12-15T21:59:13Z Failed to open asmap file from disk.
2019-12-15T21:59:13Z Error: Could not parse specified asmap: '"/path/to/testnet3/ip_asn.map"'
Error: Could not parse specified asmap: '"/path/to/testnet3/ip_asn.map"'
```

Reply:
@fjahr that version would ignore any user-specified file, e.g. case 3.

Reply:
@jonatack yeah, I am just revising my answer as I am doing more testing, commented prematurely

Reply:
I didn't think about `-asmap` call without a parameter as "use existing". But yeah, perhaps it makes sense. Especially since both of you share that intuition. Will apply your suggested change.